spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -260,6 +260,14 @@ public Lock obtain(Object lockKey) {
 			RedisLock lockInStore = this.redisTemplate.boundValueOps(this.registryKey + ":" + lockKey).get();
 			if (lockInStore == null || !lock.equals(lockInStore)) {
 				getHardThreadLocks().remove(lock);
+				try {
+					lock.unlock();
+				}
+				catch (Exception e) {
+					if (logger.isWarnEnabled()) {
+						logger.warn("Lock was released due to expiration. Will be obtained a new one...", e);
```

Comment:
logger.warn("Lock was released due to expiration. A new one will be obtained...", e);

If `useWeakReferences` is true, we'll find the same lock again below. Need to remove from that map as well under this condition.

Add `registry.setUseWeakReferences(true);` to your test and it fails.